src/request.rs

```diff
@@ -14,28 +14,113 @@ use super::SerdeValue;
 
 pub type Result<T> = std::result::Result<T, Error>;
 
+/// Parsed result of a request url with handy inspection methods.
 #[derive(Clone)]
-struct ParsedUrl(std::result::Result<Url, url::ParseError>);
+pub struct ParsedUrl {
+    /// Url parse result kept as Result since we want to defer the surfacing of it
+    /// until we do call()/send()/etc.
+    result: std::result::Result<Url, url::ParseError>,
+    origin: Option<String>,
```

Comment:
I think we don't need to carry around the original input string. It looks like the only place we currently use it is in the `.url()` accessor, so we can return the original in case it failed to parse. But I can't think of a case where it's useful to get back the original malformed URL in that case, since the request is doomed to fail anyhow. Why not define `.url()` to return "" if the input URL failed to parse?

Reply:
I just really want that original url. I feel quite strongly about it. Parsing, technically, should only happen once we do `call`, `send` etc. The only reason it surfaces sooner would be on `parsed_url()`.

Any parsed (correct?) variant is extra topping for me. Not the actual base API.

Reply:
Fair enough. In that case, what do you think of making this struct an `enum { Str(String), Url(Url), Error(url::ParseError) }`, since it's guaranteed to contain exactly one of those?

Also, I think the term `origin` should be called `original` because "origin" has a technical meaning on the web (the scheme + authority).

Reply:
> Fair enough. In that case, what do you think of making this struct an enum { Str(String), Url(Url), Error(url::ParseError) }, since it's guaranteed to contain exactly one of those?

Ok. I'll fix!

> Also, I think the term origin should be called original because "origin" has a technical meaning on the web (the scheme + authority).

Good point! Helps me to have a native speaker checking my language. I fix!

Reply:
> what do you think of making this struct an enum { Str(String), Url(Url), Error(url::ParseError) }, since it's guaranteed to contain exactly one of those?

I realized I was wrong about this! I this is guaranteed to contain one of:

1) a String _and_ a ParseError
2) a Url

So maybe `enum { Str(String, ParseError), Url(Url) }`?

Reply:
@jsha 

> So maybe `enum { Str(String, ParseError), Url(Url) }`?

I'm trying to go down this path, but tbh, it seems … awkward.

Here's what I try to do:

```rust
/// Encapsulation of the various state a url can be in.
#[derive(Clone)]
enum UrlState {
    /// String has not yet been parsed to a Url yet.
    Unparsed(Arc<String>),
    /// A successfully parsed url.
    Parsed(Url),
    /// Special state when using `Request::parsed_url().
    /// Query pairs copied out from Cow<str> in url::Url.
    ParsedAndQueryPairs(ParsedAndQueryPairs),
    /// A url that failed to parse. The string is the original we tried to parse.
    ParseError(Arc<String>, ParseError),
}

#[derive(Clone)]
struct ParsedAndQueryPairs {
    url: Url,
    query_pairs: Vec<(String, String)>,
}
```

And then try to encapsulate all state transitions neatly in a `impl UrlState`. But it's total code explosion. I've pushed it, but I think it makes things more complicated.

Reply:
Yeah, I agree this makes things way more complicated, and I preferred the previous version. Thanks for trying it out.